python/lsst/scarlet/lite/blend.py

```diff
@@ -0,0 +1,276 @@
+        ----------
+        sources:
+            The sources to fit.
+        observation:
+            The observation that contains the images,
+            PSF, etc. that are being fit.
+        """
+        self.sources = list(sources)
+        self.observation = observation
+
+        # Initialzie the iteration count and loss function
+        self.it = 0
+        self.loss: list[float] = []
+
+    @property
+    def shape(self) -> tuple[int, int, int]:
+        """Shape of the model for the entire `Blend`."""
+        return self.observation.shape
+
+    @property
+    def bbox(self) -> Box:
+        """The bounding box of the entire blend."""
+        return self.observation.bbox
+
+    @property
+    def components(self) -> list[Component]:
+        """The list of all components in the blend.
+
+        Since the list of sources might change,
+        this is always built on the fly.
+        """
+        return [c for src in self.sources for c in src.components]
+
+    def get_model(self, convolve: bool = False, use_flux: bool = False) -> Image:
+        """Generate a model of the entire blend.
+
+        Parameters
+        ----------
+        convolve:
+            Whether to convolve the model with the observed PSF in each band.
+        use_flux:
+            Whether to use the re-distributed flux associated with the source
+            instead of the component models.
+
+        Returns
+        -------
+        model:
+            The model created by combining all of the source models.
+        """
+        model = Image(
+            np.zeros(self.shape, dtype=self.observation.images.dtype),
+            bands=self.observation.bands,
+            yx0=cast(tuple[int, int], self.observation.bbox.origin[-2:]),
+        )
+
+        if use_flux:
+            for src in self.sources:
+                if src.flux is None:
+                    raise ValueError(
+                        "Some sources do not have 'flux' attribute set. Run measure.conserve_flux"
+                    )
+                src.flux.insert_into(model)
+        else:
+            for component in self.components:
+                component.get_model().insert_into(model)
+            if convolve:
+                return self.observation.convolve(model)
+        return model
+
+    def _grad_log_likelihood(self) -> Image:
+        """Gradient of the likelihood wrt the unconvolved model"""
+        model = self.get_model(convolve=True)
+        # Update the loss
+        self.loss.append(self.observation.log_likelihood(model))
+        # Calculate the gradient wrt the model d(logL)/d(model)
+        result = self.observation.weights * (model - self.observation.images)
+        result = self.observation.convolve(result, grad=True)
+        return result
+
+    @property
+    def log_likelihood(self) -> float:
+        """The current log-likelihood"""
+        if len(self.loss) == 0:
+            return self.observation.log_likelihood(self.get_model(convolve=True))
```

Comment:
`Blend.get_model` doesn't cache the model, it rebuilds it with all of the components and their _current_ parameter values.

Reply:
Oh, I see why this is confusing now and I'll add a comment to clarify. The parameters are only changed in `Blend.fit`, and in each iteration the new logL is stored in `Blend.loss`. But if the user hasn't fit the model yet, then `logL` hasn't been calculated, so we do it on the fly. In practice in the science pipelines we never do this, but it is convenient for downstream functions to always be able to calculate logL, even for an initial model.

Reply:
https://github.com/lsst/scarlet_lite/blob/tickets/DM-35562/python/lsst/scarlet/lite/blend.py#L130-L131

Technically, yes. This is python, so there are no private attributes and users can do whatever they want to internal attributes. Should I add a comment that states that this is the log likelihood for the last _fit_ model?

Reply:
Yes, something like "this is logL for the last fit model, or the current parameters if self.loss is empty" would help. I'm bringing this up because I think users might actually want to do what I described. I suppose if someone wants to manually update parameters and logL after a fit, they can set `self.loss = []`, and potentially re-fit?

Reply:
Conclusion from slack DM is to dynamically calculate logL to avoid confusion.